Support Memoizing Layers

adamgfraser opened the pull request:
Initial work on resolving #2772. The challenge is not really memoizing a layer, as we can already get a managed out of a layer and memoize that, but the change in architecture in no longer having any "global" environment and the fact that memoization gives us an effect wrapped in an effect. Currently this PR just changes the way the test environment is constructed in ZIO Test to memoize the live environment.

I am wondering if we can build on the idea of layers as "recipes" for environments to allow memoizing layers in a referentially transparent way without creating an additional effect layer. If we prevent you from just getting any managed out of a layer and make `build` equivalent to converting the recipe to the environment in some sense then I think we could potentially make `memoize` just change the "recipe" for building that layer, though that needs to be worked through more.

adamgfraser commented:
@jdegoes Initial implementation of shared layers is here. This needs some cleanup but seems to be working relatively well, with common dependencies across different layers only creating once and resources being acquired and released in the correct order.

The main issue I see right now is that since we are using layers as the keys in the `MemoMap` we need a way to has them. Right now for this demonstration I am relying on object equality but I don't think that is an acceptable solution. We probably need some kind of concept of labels when creating layers, and then we can have logic for composing those labels. It isn't great to require users to do one more thing to create a layer but relative to the previous ceremony seems manageable. Also potentially provides an easy solution to how to create resources that aren't shared, just give them a different identifier.

jdegoes commented:
@adamgfraser 

I'll review this today. High-level thoughts:

1. I think piggybacking on object equality/hash code isn't terrible, even if it does weakly violate referential transparency (weakly because for sanely-defined layers, it will have no effect on behavior, only performance, which is ordinarily exempted from referential transparency). The only place it may give unintuitive results is if someone dynamically makes a layer, e.g. `database(770, "localhost") >>> (...) ++ database(770, "localhost") >>> (...)`. But users can solve that problem by factoring out the duplication, e.g. `val db = database(770, "localhost")`. The problem with identifiers is that we can't know two layers with the same identifier actually represent the same layer; whereas we do know that two layers with the same object identity represent the same layer (since it is literally the same object). And when layers are distributed across a 2-3 million LOC application, the odds of accidental collisions is quite high, and could lead to painful and protracted debugging process.
2. I was originally thinking `shared` would be a method on `ZLayer` but after looking at this PR, in which sharing is the default, I think that makes _more_ sense. We can create `ZLayer#fresh` to make a fresh layer, if for some reason it's important to not share. The folks at Septimal Mind have not found a use case for "non-shared" common dependencies, although I can think of some (e.g. some service creates a thread pool and you need two isolated thread pools in different parts of your application for work isolation).

adamgfraser commented:
@jdegoes Great! I am most of the way through moving `ZLayer` over to the new encoding and will clean this up as part of that so will do that this morning and then your review can be more efficient.

adamgfraser commented:
@jdegoes This is ready for review.

jdegoes commented on core/shared/src/main/scala/zio/ZLayer.scala:
```diff
@@ -49,47 +61,79 @@ final case class ZLayer[-RIn, +E, +ROut <: Has[_]](value: ZManaged[RIn, E, ROut]
+   * Returns an effect that, if evaluated, will return the lazily computed
+   * result of this layer.
+   */
+  def memoize: ZManaged[RIn, Nothing, ZLayer[Any, E, ROut]] =
```
Is this useful anymore? I think we can delete it unless you have a use case in mind.

adamgfraser replied:
Yes I think that is right.

jdegoes commented on core/shared/src/main/scala/zio/ZLayer.scala:
```diff
@@ -115,13 +159,13 @@ object ZLayer {
    * Constructs a layer that purely depends on the specified service.
    */
-  def fromService[A: Tagged, E, B <: Has[_]](f: A => B): ZLayer[Has[A], E, B] =
+  def fromService[A: Tagged, B <: Has[_]](f: A => B): ZLayer[Has[A], Nothing, B] =
```
Nice cleanups

jdegoes commented on core/shared/src/main/scala/zio/ZLayer.scala:
```diff
@@ -341,4 +385,63 @@ object ZLayer {
+      finalizerRef: Ref[Exit[Any, Any] => ZIO[Any, Nothing, Any]]
+    ): ZManaged[A, E, B] =
+      ZManaged {
+        get(layer).flatMap {
```
Is this safe for arbitrary interruption, or would it be useful to add `uninterruptibleMask` to protect critical regions, and allow interruption only some places (e.g. `reservation.acquire`?)?

adamgfraser replied:
So I think we are okay here but it is definitely the right question. My analysis is:

1. First we check if the dependency is in the map and if so just return it. I don't think there is any problem with being interrupted here because we haven't acquired any resources at this point that require finalization.
2. If we don't have the resource yet, we do `layer.scope.flatMap(_.apply(self))` which creates one Managed that describes entering both the inner and outer scopes to create the dependency.
3. Then we call `reserve` on it to access the `acquire` and `release` actions of the managed separately but we still haven't actually acquired any resources.
4. The actual acquisition of resources comes in `reservation.acquire` which is in this reservation we are constructing that will create a new `Managed`, so the acquisition of the dependency and adding it to the memo map will be uninterruptible and if it starts the finalizers are guaranteed to be added to the outer scope, also uninterruptibly, because that is in the release action of the managed.

jdegoes replied:
It's tricky, but I think you're right..

jdegoes commented:
@adamgfraser 

I don't know how Lint failed in master but didn't in the PR:

<https://github.com/zio/zio/commit/5f99983b18ddc16381f3c1198d599773542eca60>

adamgfraser commented:
@jdegoes The issue is when we implemented `provideSomeLayer` in the other PR we added a variant to streams that didn't exist on this branch, so when we merged we had a compilation issue because that variant hadn't been updated to the new encoding. Anyway, it is fixed now.
